docs/denotational-design.md

```diff
@@ -1,214 +1,225 @@
+type Browser a b
+μ type Browser a b = ReaderT (Snapshot a) (State (History a) b)
+
+-- A function that will retrieve a repository given an
+-- identifier. In our case the identifier is opaque to the system.
+getRepo :: Repo -> Repo
+
+-- Find a particular History in the Repo. Again, how these things
+-- are equated and found is opaque, but we can think of them as
+-- branch or tag labels.
+getHistory :: Eq a => History a -> Repo a -> Maybe (History a)
+μ getHistory history repo =
+  find history (μ repo)
+
+-- Find if a particular artefact occurs in 0 or more histories.
+findInHistories :: a -> [History a] -> [History a]
+μ findInHistories a histories =
+  filterMaybe (findInHistory a) histories
+
+-- Find a particular artefact is in a history.
+findInHistory :: Eq a => a -> History a -> Maybe a
+μ findInHistory a history = find (== a) (μ history)
+
+-- A special Label that guarantees a starting point, i.e. ~
+root :: Label
+
+-- Get the difference between two directory views.
+diff :: Directory -> Directory -> Diff
+
+-- List the current file or directories in a given Directory view.
+listDirectory :: Directory -> NonEmpty (Label, SystemType)
+μ listDirectory directory = map f $ snd (μ directory)
+  where
+    f = \case
+      Left dir -> (fst dir, IsDirectory)
+      Right file -> (fst file, IsFile)
+
+fileName :: File -> Label
+μ fileName file = fst (μ file)
 
--- Getting the [Directory] from CommitHistory should be equivalent to building
--- the [File] and getting their Directory
-files_and_directories :: Property
-files_and_directories = ∀ history. getHistoryDirectories history ≡ map fileDirectory (getFiles (getCommits history))
+findFile :: NonEmpty Label -> Directory -> Maybe File
+μ findFile (label :| labels) directory =
+  let (label, artefacts) = (μ directory)
+  if label == label' then go labels artefacts else Nothing
+  where
+    findFileWithLabel :: Foldable f => Label -> f (Either Directory File) -> Maybe File
+    findFileWithLabel label = find (\artefact -> case artefact of
+      Left _     -> False
+      Right file -> fileLabel == label)
+
+    go :: [Label] -> [Either Directory File] -> Just File
+    go [] _ = Nothing
+    go [label] directories = findMaybe (fileWithLabel label) directories
+    go (label:labels) directories = go labels $ find ((label ==) . fst) onlyDirectories directories
+
+onlyDirectories :: Foldable f => f (Either Directory File) -> [Directory]
+onlyDirectories = filter isLeft . toList
+
+getSubDirectories :: Directory -> [Directory]
+μ getSubDirectories directory = foldMap f $ snd (μ directory)
+  where
+    f :: Either Directory File -> [Directory]
+    f = either pure []
+
+-- Definition elided
+findDirectory :: NonEmpty Label -> Directory -> Maybe Directory
+
+-- Definition elided
+fuzzyFind :: Label -> [Directory]
+
+-- A Git Snapshot is grabbing the HEAD commit of your History
+-- and turning it into a Directory
+gitSnapshot :: Snapshot [Commit]
+gitSnapshot = getDirectoryPtr . head
```

Comment:
I think this is simply:

```haskell
gitSnapshot :: GitSnapshot -- or: Snapshot Commit
gitSnapshot = ask
```

Although probably not too useful if we have `getDirectory`

Reply:
Actually:

```haskell
gitSnapshot :: Browser GitSnapshot GitSnapshot
```

Reply:
I was thinking this was more the function to pass to the `ReaderT`, so we would have `runBrowser gitSnapshot $ actions`?

Reply:
Ah yeah. I think it still doesn't quite check out because how do you take the `head` of `[a] -> b`?

Reply:
Ya, maybe it would be something like:
```haskell
maybe rootDirectory getDirectoryPtr $ Nel.head <$> nonEmpty
```

Or even saying that! We should probably have `NonEmpty a` as our set of artefacts. `Repo` should be `Maybe (NonEmpty (History a))`?

Reply:
That way the only we have `Nothing` is when our `Repo` is truly in the initial state of nothingness.

Reply:
Actually `Maybe (NonEmpty (History a))` is the same as `[History a]` :sweat_smile: